File: backend/practice.py
"""Practice test generator — analyses weak bigrams and generates
targeted word lists and bigram drills for improvement."""

import random
import sqlite3

from database import get_bigram_stats

# ─── Word bank (~400 common English words with diverse bigrams) ─────
WORD_BANK = [
    "about", "above", "across", "action", "after", "again", "against",
    "almost", "along", "already", "always", "among", "animal", "another",
    "answer", "appear", "around", "asked", "away", "back", "became",
    "because", "become", "before", "behind", "being", "below", "better",
    "between", "black", "blood", "board", "brought", "build", "built",
    "called", "came", "carry", "cause", "center", "certain", "change",
    "children", "church", "circle", "clear", "close", "cold", "common",
    "complete", "contain", "could", "country", "course", "cover", "create",
    "dark", "death", "decide", "deep", "develop", "different", "direct",
    "distant", "does", "done", "door", "down", "drive", "during", "each",
    "early", "earth", "east", "effect", "either", "enough", "enter",
    "equal", "even", "ever", "every", "example", "experience", "face",
    "fact", "fall", "family", "father", "feel", "feet", "field", "figure",
    "final", "find", "fine", "fire", "first", "five", "floor", "follow",
    "food", "foot", "force", "form", "found", "four", "free", "friend",
    "from", "front", "full", "game", "gave", "girl", "give", "glass",
    "going", "gold", "gone", "good", "great", "green", "ground", "group",
    "grow", "half", "hand", "happen", "hard", "have", "head", "hear",
    "heart", "heavy", "help", "here", "high", "hold", "home", "horse",
    "hour", "house", "however", "human", "hundred", "idea", "important",
    "include", "inside", "instead", "interest", "into", "island", "just",
    "keep", "kind", "knew", "know", "known", "land", "language", "large",
    "last", "later", "learn", "least", "leave", "less", "letter", "life",
    "light", "like", "line", "list", "little", "live", "long", "look",
    "love", "made", "make", "many", "matter", "mean", "might", "mind",
    "money", "month", "more", "morning", "most", "mother", "mountain",
    "move", "much", "music", "must", "name", "near", "need", "never",
    "new", "next", "night", "north", "note", "nothing", "notice", "number",
    "object", "ocean", "once", "only", "open", "order", "other", "our",
    "outside", "over", "own", "page", "paper", "part", "pass", "past",
    "pattern", "people", "perhaps", "period", "person", "picture", "piece",
    "place", "plain", "plan", "plane", "plant", "play", "point", "pose",
    "power", "problem", "produce", "product", "pull", "put", "question",
    "quick", "rain", "reach", "read", "ready", "really", "record", "red",
    "remember", "rest", "right", "river", "road", "rock", "room", "round",
    "rule", "said", "same", "saw", "school", "second", "section", "seen",
    "serve", "several", "shape", "ship", "short", "should", "show", "side",
    "simple", "since", "sing", "slow", "small", "snow", "something",
    "song", "soon", "south", "space", "special", "stand", "star", "start",
    "state", "stay", "step", "still", "stood", "stop", "story", "street",
    "strong", "study", "such", "surface", "system", "table", "take",
    "talk", "teach", "tell", "than", "that", "their", "them", "then",
    "there", "these", "they", "thing", "think", "this", "those", "though",
    "thought", "three", "through", "time", "together", "told", "took",
    "toward", "tree", "true", "try", "turn", "under", "understand",
    "until", "upon", "used", "using", "very", "voice", "wait", "walk",
    "want", "war", "watch", "water", "week", "weight", "well", "went",
    "were", "west", "what", "wheel", "when", "where", "which", "while",
    "white", "whole", "will", "wind", "with", "within", "without",
    "woman", "wonder", "word", "work", "world", "would", "write", "year",
    "young", "your",
]


def extract_bigrams(word: str) -> list[str]:
    """Extract all consecutive 2-character pairs from a word.

    Example: "there" → ["th", "he", "er", "re"]
    """
    return [word[i] + word[i + 1] for i in range(len(word) - 1)]


def score_word(word: str, weak_bigrams: dict, worst_bigram: str) -> float:
    """Score a word based on how many weak bigrams it contains.

    Each matching weak bigram contributes its error_rate to the score.
    The worst bigram gets a 1.5× weighting boost.
    Multiple distinct weak bigrams in one word get a 1.2× bonus per extra match.

    Args:
        word: The word to score.
        weak_bigrams: Dict mapping bigram → error_rate.
        worst_bigram: The single worst bigram (for weighting boost).

    Returns:
        A float score (higher = better practice word).
    """
    word_bigrams = extract_bigrams(word)
    score = 0.0
    matches = 0

    for bg in word_bigrams:
        if bg in weak_bigrams:
            weight = weak_bigrams[bg]
            if bg == worst_bigram:
                weight *= 1.5  # boost for the worst bigram
            score += weight
            matches += 1

    # Bonus for words containing multiple weak bigrams
    if matches > 1:
        score *= 1.0 + (matches - 1) * 0.2

    return score


def generate_drill(weak_bigrams: list[dict], reps: int = 5) -> str:
    """Generate a pure bigram drill string for muscle memory.

    Example: "er er er er er  in in in in in  th th th th th"

    Args:
        weak_bigrams: List of bigram dicts with at least "bigram" key.
        reps: How many times to repeat each bigram.

    Returns:
        A space-separated drill string.
    """
    parts = []
    for bg_info in weak_bigrams:
        bg = bg_info["bigram"]
        parts.extend([bg] * reps)
    return "  ".join(parts)  # double-space between different bigrams


def generate_practice(
    db: sqlite3.Connection,
    count: int = 10,
    word_count: int = 35,
) -> dict:
    """Generate a targeted practice test based on the user's weakest bigrams.

    Args:
        db: An open SQLite connection.
        count: Number of weak bigrams to target (default 10).
        word_count: Number of words in the generated test (default 35).

    Returns:
        A dict with targeted_bigrams, practice_words, drill_text, and metadata.
    """
    # ── Step 1: Get weak bigrams ──────────────────────────────────
    bigram_stats = get_bigram_stats(db)
    all_bigrams = bigram_stats.get("bigrams", [])
    total_sessions = bigram_stats.get("total_sessions", 0)

    # No sessions yet
    if total_sessions == 0:
        return {
            "error": "Complete at least one typing test first",
            "practice_words": [],
            "drill_text": "",
            "total_words": 0,
            "total_bigrams_targeted": 0,
            "targeted_bigrams": [],
        }

    # Filter to bigrams with error_rate > 0 (we only care about mistakes)
    weak = [b for b in all_bigrams if b["error_rate"] > 0]

    if len(weak) < 3:
        return {
            "warning": (
                f"Need more typing data. Only found {len(weak)} bigrams "
                "with errors. Keep practicing!"
            ),
            "targeted_bigrams": weak,
            "practice_words": _fallback_words(word_count),
            "drill_text": generate_drill(weak) if weak else "",
            "total_words": word_count,
            "total_bigrams_targeted": len(weak),
        }

    # Take top N weak bigrams
    targeted = weak[:count]

    # ── Step 2: Build lookup dicts ─────────────────────────────────
    weak_dict = {b["bigram"]: b["error_rate"] for b in targeted}
    worst = targeted[0]["bigram"]  # already sorted by error_rate desc

    # ── Step 3: Score all words ───────────────────────────────────
    scored = []
    for word in WORD_BANK:
        s = score_word(word, weak_dict, worst)
        if s > 0:
            scored.append((word, s))

    # Sort by score descending
    scored.sort(key=lambda x: x[1], reverse=True)

    # ── Step 4: Select words for the test ─────────────────────────
    practice_words = []
    if scored:
        # Take top-scoring words, allow repetition of very high scorers
        top_words = [w for w, _ in scored]
        idx = 0
        while len(practice_words) < word_count:
            practice_words.append(top_words[idx % len(top_words)])
            idx += 1
    else:
        practice_words = _fallback_words(word_count)

    # Slight shuffle so the test doesn't feel too mechanical
    random.shuffle(practice_words)

    # ── Step 5: Generate drill ────────────────────────────────────
    drill_text = generate_drill(targeted)

    return {
        "targeted_bigrams": [
            {"bigram": b["bigram"], "error_rate": b["error_rate"]}
            for b in targeted
        ],
        "practice_words": practice_words,
        "drill_text": drill_text,
        "total_words": len(practice_words),
        "total_bigrams_targeted": len(targeted),
    }


def _fallback_words(count: int) -> list[str]:
    """Return random words from the bank when not enough data exists."""
    return random.sample(
        WORD_BANK,
        min(count, len(WORD_BANK)),
    )

File: backend/practice.js
// This list is deliberately duplicated rather than shared: Railway builds each
// service from its own directory, so a module at the repo root would not be in
// the backend build context. The frontend keeps its own copy in
// frontend/src/components/TypingTest.jsx. `node scripts/check-word-bank.mjs`
// fails if the two copies drift apart.
export const WORD_BANK = [
  'a', 'able', 'about', 'above', 'across', 'action', 'add', 'after', 'again', 'against',
  'ago', 'air', 'all', 'almost', 'along', 'already', 'also', 'always', 'among', 'an',
  'and', 'animal', 'another', 'answer', 'any', 'appear', 'area', 'around', 'as', 'ask',
  'asked', 'at', 'away', 'back', 'be', 'beauty', 'became', 'because', 'become', 'before',
  'begin', 'behind', 'being', 'below', 'best', 'better', 'between', 'big', 'bird', 'black',
  'blood', 'blue', 'board', 'boat', 'body', 'book', 'both', 'box', 'bring', 'brought',
  'build', 'built', 'busy', 'but', 'by', 'called', 'came', 'can', 'car', 'care',
  'carry', 'cause', 'center', 'certain', 'change', 'check', 'children', 'church', 'circle', 'city',
  'class', 'clear', 'close', 'cold', 'come', 'common', 'complete', 'contain', 'correct', 'could',
  'country', 'course', 'cover', 'create', 'cry', 'dark', 'day', 'death', 'decide', 'deep',
  'develop', 'different', 'direct', 'distant', 'do', 'does', 'dog', 'done', 'door', 'down',
  'drive', 'dry', 'during', 'each', 'early', 'earth', 'ease', 'east', 'effect', 'either',
  'end', 'enough', 'enter', 'equal', 'equate', 'even', 'ever', 'every', 'example', 'experience',
  'eyes', 'face', 'fact', 'fall', 'family', 'far', 'fast', 'father', 'feel', 'feet',
  'few', 'field', 'figure', 'fill', 'final', 'find', 'fine', 'fire', 'first', 'five',
  'floor', 'fly', 'follow', 'food', 'foot', 'for', 'force', 'form', 'found', 'four',
  'free', 'friend', 'from', 'front', 'full', 'game', 'gave', 'get', 'girl', 'give',
  'glass', 'go', 'going', 'gold', 'gone', 'good', 'got', 'govern', 'great', 'green',
  'ground', 'group', 'grow', 'half', 'hand', 'happen', 'hard', 'have', 'he', 'head',
  'hear', 'heard', 'heart', 'heat', 'heavy', 'help', 'her', 'here', 'high', 'him',
  'his', 'hold', 'home', 'horse', 'hot', 'hour', 'house', 'how', 'however', 'human',
  'hundred', 'idea', 'if', 'important', 'in', 'inch', 'include', 'inside', 'instead', 'interest',
  'into', 'island', 'it', 'its', 'just', 'keep', 'kind', 'king', 'knew', 'know',
  'known', 'land', 'language', 'large', 'last', 'later', 'laugh', 'lay', 'lead', 'learn',
  'least', 'leave', 'left', 'less', 'letter', 'life', 'light', 'like', 'line', 'list',
  'listen', 'little', 'live', 'long', 'look', 'love', 'machine', 'made', 'make', 'many',
  'map', 'mark', 'matter', 'me', 'mean', 'measure', 'might', 'mile', 'mind', 'minute',
  'miss', 'money', 'month', 'moon', 'more', 'morning', 'most', 'mother', 'mountain', 'move',
  'much', 'multiply', 'music', 'must', 'my', 'name', 'near', 'need', 'never', 'new',
  'next', 'night', 'no', 'north', 'not', 'note', 'nothing', 'notice', 'noun', 'now',
  'number', 'object', 'ocean', 'of', 'often', 'oh', 'old', 'on', 'once', 'one',
  'only', 'open', 'or', 'order', 'other', 'our', 'out', 'outside', 'over', 'own',
  'page', 'paint', 'paper', 'part', 'pass', 'past', 'pattern', 'people', 'perhaps', 'period',
  'person', 'picture', 'piece', 'place', 'plain', 'plan', 'plane', 'plant', 'play', 'point',
  'pose', 'possible', 'pound', 'power', 'problem', 'produce', 'product', 'pull', 'put', 'question',
  'quick', 'rain', 'ran', 'reach', 'read', 'ready', 'really', 'record', 'red', 'remember',
  'rest', 'right', 'river', 'road', 'rock', 'room', 'round', 'rule', 'run', 'said',
  'same', 'saw', 'say', 'school', 'second', 'section', 'see', 'seem', 'seen', 'serve',
  'set', 'several', 'shape', 'she', 'ship', 'short', 'should', 'show', 'side', 'simple',
  'since', 'sing', 'six', 'slow', 'small', 'snow', 'so', 'some', 'something', 'song',
  'soon', 'south', 'space', 'special', 'spell', 'stand', 'star', 'start', 'state', 'stay',
  'stead', 'step', 'still', 'stood', 'stop', 'story', 'street', 'strong', 'study', 'such',
  'surface', 'system', 'table', 'tail', 'take', 'talk', 'teach', 'tell', 'ten', 'test',
  'than', 'that', 'the', 'their', 'them', 'then', 'there', 'these', 'they', 'thing',
  'think', 'this', 'those', 'though', 'thought', 'thousand', 'three', 'through', 'time', 'tire',
  'to', 'together', 'told', 'took', 'top', 'toward', 'town', 'travel', 'tree', 'true',
  'try', 'turn', 'two', 'under', 'understand', 'unit', 'until', 'up', 'upon', 'us',
  'use', 'used', 'using', 'usual', 'verb', 'very', 'voice', 'vowel', 'wait', 'walk',
  'want', 'war', 'warm', 'watch', 'water', 'way', 'we', 'week', 'weight', 'well',
  'went', 'were', 'west', 'what', 'wheel', 'when', 'where', 'which', 'while', 'white',
  'who', 'whole', 'will', 'wind', 'with', 'within', 'without', 'woman', 'wonder', 'word',
  'work', 'world', 'would', 'write', 'year', 'yes', 'you', 'young', 'your',
];

export function extractBigrams(word) {
  const bigrams = [];
  for (let i = 0; i < word.length - 1; i++) {
    bigrams.push(word[i] + word[i + 1]);
  }
  return bigrams;
}

export function scoreWord(word, weakBigrams, worstBigram) {
  const bigrams = extractBigrams(word);
  let score = 0.0;
  let matches = 0;

  for (const bg of bigrams) {
    if (bg in weakBigrams) {
      let weight = weakBigrams[bg];
      if (bg === worstBigram) weight *= 1.5;
      score += weight;
      matches += 1;
    }
  }

  if (matches > 1) {
    score *= 1.0 + (matches - 1) * 0.2;
  }
  return score;
}

export function generateDrill(weakBigrams, reps = 5) {
  const parts = [];
  for (const bgInfo of weakBigrams) {
    for (let i = 0; i < reps; i++) {
      parts.push(bgInfo.bigram);
    }
  }
  return parts.join('  ');
}

// Every bigram that appears in at least one word-bank word. Used to discard
// targets the bank cannot produce — e.g. "Th" or "d," from quotes mode. Such
// bigrams score no word at all, so they only waste slots in the target list.
const BANK_BIGRAMS = new Set();
for (const word of WORD_BANK) {
  for (const bg of extractBigrams(word)) BANK_BIGRAMS.add(bg);
}

// Fisher-Yates. Array.sort(() => 0.5 - Math.random()) is biased and produces
// an uneven distribution.
export function shuffle(items) {
  const out = [...items];
  for (let i = out.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [out[i], out[j]] = [out[j], out[i]];
  }
  return out;
}

function fallbackWords(count) {
  return shuffle(WORD_BANK).slice(0, Math.min(count, WORD_BANK.length));
}

// Pure: the caller fetches the user's bigram stats and passes them in, so this
// module has no database dependency and can be tested with a plain object.
export function generatePractice({ stats, count = 10, wordCount = 35 }) {
  const allBigrams = stats?.bigrams || [];
  const totalSessions = stats?.total_sessions || 0;

  if (totalSessions === 0) {
    return {
      error: "Complete at least one typing test first",
      practice_words: [],
      drill_text: "",
      total_words: 0,
      total_bigrams_targeted: 0,
      targeted_bigrams: []
    };
  }

  const weak = allBigrams.filter((b) => b.error_rate > 0);

  if (weak.length < 3) {
    return {
      warning: `Need more typing data. Only found ${weak.length} bigrams with errors. Keep practicing!`,
      targeted_bigrams: weak,
      practice_words: fallbackWords(wordCount),
      drill_text: weak.length > 0 ? generateDrill(weak) : "",
      total_words: wordCount,
      total_bigrams_targeted: weak.length
    };
  }

  // Only target bigrams the word bank can actually produce. Stats include
  // quote-mode bigrams with capitals and punctuation ("Th", "d,") that no
  // lowercase bank word contains; targeting those crowds out usable targets.
  const candidates = weak.filter((b) => BANK_BIGRAMS.has(b.bigram.toLowerCase()));

  // Dedupe case-insensitively ("Th" vs "th"), keeping the worst first.
  const seen = new Set();
  const targeted = [];
  for (const b of (candidates.length > 0 ? candidates : weak)) {
    const key = b.bigram.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    targeted.push(b);
    if (targeted.length >= count) break;
  }

  // Keyed lowercase so bank words (all lowercase) can match.
  const weakDict = Object.fromEntries(
    targeted.map((b) => [b.bigram.toLowerCase(), b.error_rate])
  );
  const worst = targeted[0].bigram.toLowerCase();

  const scored = [];
  for (const word of WORD_BANK) {
    const s = scoreWord(word, weakDict, worst);
    if (s > 0) scored.push({ word, score: s });
  }
  scored.sort((a, b) => b.score - a.score);

  // Build a pool of DISTINCT words — strongest scorers first, then topped up
  // from the word bank. The top-up is essential: if only one bank word matches
  // the targeted bigrams, cycling a bare candidate list would repeat that
  // single word for the whole test.
  const MAX_STRONG = 24;
  const pool = scored.slice(0, MAX_STRONG).map((item) => item.word);

  if (pool.length < wordCount) {
    const chosen = new Set(pool);
    // Prefer other bank words that also exercise a targeted bigram.
    pool.push(
      ...shuffle(
        WORD_BANK.filter(
          (w) => !chosen.has(w) && extractBigrams(w).some((bg) => bg in weakDict)
        )
      )
    );

    // Still short? Top up with any remaining word-bank words.
    if (pool.length < wordCount) {
      const used = new Set(pool);
      pool.push(...shuffle(WORD_BANK.filter((w) => !used.has(w))));
    }
  }

  // Each word appears once unless the bank is smaller than the requested count
  // (impossible with the current word bank and a max of 100 words).
  const practiceWords = [];
  for (let i = 0; i < wordCount; i++) {
    practiceWords.push(pool[i % pool.length]);
  }

  return {
    targeted_bigrams: targeted.map((b) => ({ bigram: b.bigram, error_rate: b.error_rate })),
    practice_words: shuffle(practiceWords),
    drill_text: generateDrill(targeted),
    total_words: practiceWords.length,
    total_bigrams_targeted: targeted.length
  };
}